src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -1438,185 +1449,169 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         place: &Place<'tcx>,
         including_downcast: IncludingDowncast,
     ) -> Option<String> {
-        let mut buf = String::new();
-        match self.append_place_to_string(place, &mut buf, false, &including_downcast) {
-            Ok(()) => Some(buf),
-            Err(()) => None,
-        }
+        let neo_place = self.infcx.tcx.as_new_place(place);
+        self.append_place_to_string(&neo_place, &including_downcast).ok()
     }
 
     /// Appends end-user visible description of `place` to `buf`.
     fn append_place_to_string(
         &self,
-        place: &Place<'tcx>,
-        buf: &mut String,
-        mut autoderef: bool,
+        place: &NeoPlace<'tcx>,
         including_downcast: &IncludingDowncast,
-    ) -> Result<(), ()> {
-        match *place {
-            Place::Promoted(_) => {
-                buf.push_str("promoted");
-            }
-            Place::Local(local) => {
-                self.append_local_to_string(local, buf)?;
-            }
-            Place::Static(ref static_) => {
-                buf.push_str(&self.infcx.tcx.item_name(static_.def_id).to_string());
+    ) -> Result<String, ()> {
+        debug!("describe_place: {:?}", place);
+
+        let mut place_str = String::new();
+        debug!("describe_place-base: {:?}", &place.base);
+        let no_local_name = if self.mir.local_decls.is_empty() {
```

Comment:
This logic is confusing and doesn't quite seem to map to what we had before. I would expect something like

```rust
fn place_base_to_string(
    &self, 
    base: &NeoPlaceBase<'tcx>,
) -> Result<String, ()> {
    match base {
      Promoted => Ok("promoted"),
      Local(local) => self.local_str(local),
      Static(..) => ..,
  }
}
```

and then in this function

```rust
let mut place_str = self.place_base_to_string(place.base);
for proj in place.projections {
    ...
}
```